src/kbmod/search/kernels.cu

```diff
@@ -19,11 +19,62 @@
 
 #include "common.h"
 #include "cuda_errors.h"
-
-//#include "image_stack.h"
+#include "psi_phi_array_ds.h"
 
 namespace search {
 
+extern "C" void device_allocate_psi_phi_array(PsiPhiArray* data) {
+    if (!data->cpu_array_allocated())
+        throw std::runtime_error("CPU data is not allocated.");
+    if (data->gpu_array_allocated())
+        throw std::runtime_error("GPU data is already allocated.");
+
+    void* device_array_ptr;
+    checkCudaErrors(cudaMalloc((void **)&device_array_ptr, data->get_total_array_size()));
+    checkCudaErrors(cudaMemcpy(device_array_ptr,
+                               data->get_cpu_array_ptr(),
+                               data->get_total_array_size(),
+                               cudaMemcpyHostToDevice));
+    data->set_gpu_array_ptr(device_array_ptr);
+}
```

Comment:
this is the function that allocated and sets the array on the gpu device memory, and we enforce in the search that the `gpu_array_ptr` has been set, but it's not actually been used/read in at the moment right? so during runtime we just allocate the array memory, then pass the data in again for the process? (this is fine for the initial changes and if we'll actually read the pre-loaded array later I just wanted to make sure I understand)

Reply:
It is used in the search. The `PsiPhiArray` is passed into the search function `deviceSearchFilter` and the pointer to the GPU data is extracted in line 315:

```
searchFilterImages<<<blocks, threads>>>(psi_phi_array.get_meta_data(), psi_phi_array.get_gpu_array_ptr(), device_img_times,
                                            params, num_trajectories, device_tests, device_search_results);
```